src/devices/Ak8963/Ak8963.cs

```diff
@@ -0,0 +1,275 @@
+            // When powering the AK893, doc says 50 ms needed
+            Thread.Sleep(50);
+        }
+
+        /// <summary>
+        /// Get the device information
+        /// </summary>
+        /// <returns></returns>
+        public byte GetDeviceInfo()
+        {
+            return ReadByte(Register.INFO);
+        }
+
+        /// <summary>
+        /// Get the magnetometer bias
+        /// </summary>
+        public Vector3 MagnometerBias
+        {
+            get { return _magnetometerBias; }
+            set { _magnetometerBias = value; }
+        }
+
+        /// <summary>
+        /// Calibrate the magnetometer. Make sure your sensor is as far as possible of magnet
+        /// Calculate as well the magnetometer bias
+        /// </summary>
+        /// <returns>Returns the factory calibration data</returns>
+        public Vector3 CalibrateMagnetometer()
+        {
+            Vector3 calib = new Vector3();
+            Span<byte> rawData = stackalloc byte[3];
+
+            var oldPower = MeasurementMode;
+
+            // Stop the magetometer
+            MeasurementMode = MeasurementMode.PowerDown;
+            // Enter the magnetometer Fuse mode to read the calibration data
+            MeasurementMode = MeasurementMode.FuseRomAccess;
+            // Read the data
+            ReadByteArray(Register.ASAX, rawData);
+            calib.X = (float)((rawData[0] - 128) / 256.0 + 1.0);
+            calib.Y = (float)((rawData[1] - 128) / 256.0 + 1.0);
+            calib.Z = (float)((rawData[2] - 128) / 256.0 + 1.0);
+            MeasurementMode = MeasurementMode.PowerDown;
+            MeasurementMode = oldPower;
+
+            // Now calculate the bias
+            // Store old mode to restore after
+            var oldMode = MeasurementMode;
+            Vector3 minbias = new Vector3();
+            Vector3 maxbias = new Vector3();
+            int numberMeasurements = 1500;
+
+            // Setup the 100Hz continous mode
+            MeasurementMode = MeasurementMode.ContinousMeasurement100Hz;
+            for (int reading = 0; reading < numberMeasurements; reading++)
+            {
+                var bias = ReadMagnetometer(true);
+                minbias.X = Math.Min(bias.X, minbias.X);
+                minbias.Y = Math.Min(bias.Y, minbias.Y);
+                minbias.Z = Math.Min(bias.Z, minbias.Z);
+                maxbias.X = Math.Max(bias.X, maxbias.X);
+                maxbias.Y = Math.Max(bias.Y, maxbias.Y);
+                maxbias.Z = Math.Max(bias.Z, maxbias.Z);
+                // 10 ms = 100Hz, so waiting to make sure we have new data
+                Thread.Sleep(10);
+            }
+            // Store the bias
+            _magnetometerBias = ((maxbias + minbias) / 2) + calib;
+
+            return calib;
+        }
+
+        /// <summary>
+        /// True if there is a data to read
+        /// </summary>
+        public bool WaitForDataReady => (ReadByte(Register.ST1) & 0x01) == 0x01;
```

Comment:
NIT: Should we add a non-blocking WaitForDataReadyAsync?

Reply:
That can be implemented at the user code level. This property returns true if there is any data to read. So in a higher level code, you can decide to do a blocking loop or a loop where you'll put a timer or anything like this. So I'd rather let the user decide how he/she wants to handle this

Reply:
Looks like this code is not blocking, so I think that the issue then is the naming is incorrect. This should be something like `HasDataToRead` or something like that. Wait on the name means that this would be a blocking call and that it would basically spin and block until there is data available, which is why I originally proposed to have an Async method. Given that this isn't really waiting, I agree that we shouldn't have an async version, but the naming is wrong.